codex/sales/reservations.nim

```diff
@@ -429,47 +458,51 @@ proc returnBytesToAvailability*(
-    return failure(error)
+    without var reservation =? (await self.get(key, Reservation)), error:
+      return failure(error)
 
-  # We are ignoring bytes that are still present in the Reservation because
-  # they will be returned to Availability through `deleteReservation`.
-  let bytesToBeReturned = bytes - reservation.size
+    # We are ignoring bytes that are still present in the Reservation because
+    # they will be returned to Availability through `deleteReservation`.
+    let bytesToBeReturned = bytes - reservation.size
 
-  if bytesToBeReturned == 0:
-    trace "No bytes are returned", requestSizeBytes = bytes, returningBytes = bytesToBeReturned
-    return success()
+    if bytesToBeReturned == 0:
+      trace "No bytes are returned", requestSizeBytes = bytes, returningBytes = bytesToBeReturned
+      return success()
 
-  trace "Returning bytes", requestSizeBytes = bytes, returningBytes = bytesToBeReturned
+    trace "Returning bytes", requestSizeBytes = bytes, returningBytes = bytesToBeReturned
 
-  # First lets see if we can re-reserve the bytes, if the Repo's quota
-  # is depleted then we will fail-fast as there is nothing to be done atm.
-  if reserveErr =? (await self.repo.reserve(bytesToBeReturned.truncate(uint))).errorOption:
-    return failure(reserveErr.toErr(ReserveFailedError))
+    # First lets see if we can re-reserve the bytes, if the Repo's quota
+    # is depleted then we will fail-fast as there is nothing to be done atm.
+    if reserveErr =? (await self.repo.reserve(bytesToBeReturned.truncate(uint))).errorOption:
+      return failure(reserveErr.toErr(ReserveFailedError))
 
-  without availabilityKey =? availabilityId.key, error:
-    return failure(error)
+    without availabilityKey =? availabilityId.key, error:
+      return failure(error)
 
-  without var availability =? await self.get(availabilityKey, Availability), error:
-    return failure(error)
+    without var availability =? await self.get(availabilityKey, Availability), error:
+      return failure(error)
 
-  availability.freeSize += bytesToBeReturned
+    availability.freeSize += bytesToBeReturned
 
-  # Update availability with returned size
-  if updateErr =? (await self.update(availability)).errorOption:
+    # Update availability with returned size
+    if updateErr =? (await self.update(availability)).errorOption:
 
-    trace "Rolling back returning bytes"
-    if rollbackErr =? (await self.repo.release(bytesToBeReturned.truncate(uint))).errorOption:
-      rollbackErr.parent = updateErr
-      return failure(rollbackErr)
+      trace "Rolling back returning bytes"
+      if rollbackErr =? (await self.repo.release(bytesToBeReturned.truncate(uint))).errorOption:
+        rollbackErr.parent = updateErr
+        return failure(rollbackErr)
 
-    return failure(updateErr)
+      return failure(updateErr)
 
-  return success()
+    return success()
+  finally:
+    if lock.locked():
+      lock.release()
 
 proc release*(
```

Comment:
I suspect `release` should also have a lock. It seems then every function will have a lock if they do this flow:
- Read + deserialize
- Modify object
- Serialize and write/update

Reply:
Yes, but `release` is about Reservation and not Availabilities. The locks I am introducing here focus on locking for changes of sizes in Availabilities as that is the main thing that is changing. Reservation is always locked already to one Storage Request and one context, there should not be any concurrence here.